test/server/cards/01_SOR/units/GuardianOfTheWhills.spec.ts

```diff
@@ -0,0 +1,143 @@
+    integration(function (contextRef) {
+        describe('Guardian of the Whills\' ability', function () {
+            it('should decrease the cost of the first upgrade played on it by 1 resource, once per round', function () {
+                contextRef.setupTest({
+                    phase: 'action',
+                    player1: {
+                        hand: ['jetpack', 'devotion', 'electrostaff', 'foundling', 'entrenched', 'jedi-lightsaber', 'mandalorian-armor',
+                            'vambrace-grappleshot', 'survivors-gauntlet', 'moisture-farmer', 'protector', 'razor-crest#reliable-gunship'],
+                        groundArena: ['guardian-of-the-whills'],
+                        leader: 'chewbacca#walking-carpet', // vigilance aspect
+                        base: 'chopper-base'
+                    },
+                    player2: {
+                        hand: ['waylay', 'confiscate'],
+                        spaceArena: ['system-patrol-craft'],
+                        leader: 'chewbacca#walking-carpet', // vigilance aspect
+                    }
+                });
+
+                const { context } = contextRef;
+
+                // The first upgrade put on the whills should be 1 cheaper
+                context.player1.clickCard(context.jetpack);
+                expect(context.player1).toBeAbleToSelectExactly([context.guardianOfTheWhills, context.systemPatrolCraft]);
+                context.player1.clickCard(context.guardianOfTheWhills);
+                expect(context.guardianOfTheWhills).toHaveExactUpgradeNames([context.jetpack.internalName]);
+                expect(context.player1.exhaustedResourceCount).toBe(1);
+
+                context.player2.passAction();
+
+                // Any further upgrades are full price
+                context.player1.clickCard(context.devotion);
+                expect(context.player1).toBeAbleToSelectExactly([context.guardianOfTheWhills, context.systemPatrolCraft]);
+                context.player1.clickCard(context.guardianOfTheWhills);
+                expect(context.guardianOfTheWhills).toHaveExactUpgradeNames([context.jetpack.internalName, context.devotion.internalName]);
+                expect(context.player1.exhaustedResourceCount).toBe(3);
+
+                // Test round ending resets limit
+                context.moveToNextActionPhase();
+                context.player1.clickCard(context.electrostaff);
+                expect(context.player1).toBeAbleToSelectExactly([context.guardianOfTheWhills, context.systemPatrolCraft]);
+                context.player1.clickCard(context.guardianOfTheWhills);
+                expect(context.guardianOfTheWhills).toHaveExactUpgradeNames([context.jetpack.internalName, context.devotion.internalName, context.electrostaff.internalName]);
+                expect(context.player1.exhaustedResourceCount).toBe(1);
+
+                // Test this doesn't accidentally decrease the cost playing on non-Whills units
```

Comment:
I think this test would be better as the first test in this action phase since the limit is reached already at this point

Reply:
Are you talking about the test starting on line 47? we move to next action phase first?

Reply:
Oh pfft missed that